src/appleseed/renderer/modeling/bsdf/phasefunctionbsdf.h

```diff
@@ -0,0 +1,164 @@
+
+//
+// This source file is part of appleseed.
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2018 Artem Bishev, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#pragma once
+
+// Interface header.
+#include "specularbtdf.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/lighting/scatteringmode.h"
+#include "renderer/kernel/shading/directshadingcomponents.h"
+#include "renderer/modeling/bsdf/bsdf.h"
+#include "renderer/modeling/bsdf/bsdfwrapper.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/basis.h"
+#include "foundation/math/fresnel.h"
+#include "foundation/math/phasefunction.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/containers/dictionary.h"
+
+// Standard headers.
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+
+namespace renderer
+{
+    //
+    // Phase Function BSDF.
+    //
+
+    APPLESEED_DECLARE_INPUT_VALUES(PhaseFunctionBSDFInputValues)
+    {
+        Spectrum m_albedo;
+    };
+
+    class PhaseFunctionBSDF
+        : public BSDF
+    {
+    public:
+        std::unique_ptr<foundation::PhaseFunction> m_phase_function;
+
+        PhaseFunctionBSDF(
+            const char*                 name,
+            const ParamArray&           params)
+            : BSDF(name, AllBSDFTypes, ScatteringMode::Volume, params)
+        {
+            m_inputs.declare("scattering_coefficient", InputFormatSpectralReflectance);
+        }
+
+        void release() override
+        {
+            delete this;
+        }
+
+        const char* get_model() const override
+        {
+            return "phasefunction_bsdf";
```

Comment:
We're not entirely consistent in how we name models (e.g. we have both `orennayar_brdf` and `better_dipole_bssrdf`); that said in the majority of cases we use underscores to separate words, so I would suggest `phase_function_bsdf` here.